devito/builtins.py

```diff
@@ -53,6 +99,165 @@ def smooth(f, g, axis=None):
+            f[slices] = g.data[tuple(indices)]
+        elif isinstance(f, dv.Function):
+            f.data[slices] = g.data[tuple(indices)]
+        else:
+            raise NotImplementedError
+
+    lw = int(_order*sigma + 0.5)
+
+    # Create the padded grid:
+    objective_domain = ObjectiveDomain(lw)
+    try:
+        shape_padded = np.array(f.grid.shape) + 2*lw
+    except AttributeError:
+        shape_padded = np.array(f.shape) + 2*lw
+    grid = dv.Grid(shape=shape_padded, subdomains=objective_domain)
+
+    f_c = dv.Function(name='f_c', grid=grid, space_order=2*lw,
+                      coefficients='symbolic', dtype=np.int32)
+    f_o = dv.Function(name='f_o', grid=grid, coefficients='symbolic', dtype=np.int32)
+
+    weights = np.exp(-0.5/sigma**2*(np.linspace(-lw, lw, 2*lw+1))**2)
+    weights = weights/weights.sum()
+
+    additional_expressions = {}
+    for d in f_c.dimensions:
+        expr = {}
+        lhs = []
+        rhs = []
+        options = []
+
+        lhs.append(f_o)
+        rhs.append(dv.generic_derivative(f_c, d, 2*lw, 1))
+        coeffs = dv.Coefficient(1, f_c, d, weights)
+        options.append({'coefficients': dv.Substitutions(coeffs),
+                        'subdomain': grid.subdomains['objective_domain']})
+        lhs.append(f_c)
+        rhs.append(f_o)
+        options.append({'subdomain': grid.subdomains['objective_domain']})
+
+        expr['lhs'] = lhs
+        expr['rhs'] = rhs
+        expr['options'] = options
+        additional_expressions[d] = expr
+
+    initialize_function(f_c, f.data[:], lw,
+                        additional_expressions=additional_expressions,
+                        mode='reflect', name='smooth')
+
+    fset(f, f_c)
+
+    return f
+
+
+def initialize_function(function, data, nbl, additional_expressions=dict(),
+                        mode='constant', name='padfunc'):
+    """
+    Initialize a `Function` with the given ``data``. ``data``
+    does *not* include the ``nbl`` outer/boundary layers; these are added via padding
+    by this function.
+
+    Parameters
+    ----------
+    function : Function
+        The initialised object.
+    data : ndarray of Function
+        The data used for initialisation.
+    nbl : int
+        Number of outer layers (such as PML layers for boundary damping).
+    additional_expressions : dict, optional
+        Dictionary containing, for each dimension of function, a sub-dictionary
+        containing the following keys:
+        1) 'lhs': List of additional expressions to be added to the LHS expressions list.
+        2) 'rhs': List of additional expressions to be added to the RHS expressions list.
+        3) 'options': Options pertaining to the additional equations that will be
+        constructed.
+    mode : str, optional
+        The function initialisation mode. 'constant' and 'reflect' are
+        accepted.
+    name : str, optional
+        The name assigned to the operator.
+    """
```

Comment:
any chance we can add example here, showing the flexibility given by ``additional_expressions`` ?